libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/FreeDriveLocationUpdater.java

```diff
@@ -0,0 +1,148 @@
+  private final NavigationEventDispatcher navigationEventDispatcher;
+  private final MapboxNavigator mapboxNavigator;
+  private final ScheduledExecutorService executorService;
+  private final LocationEngineCallback<LocationEngineResult> callback = new CurrentLocationEngineCallback(this);
+  private ScheduledFuture future;
+  private Location rawLocation;
+
+  FreeDriveLocationUpdater(@NonNull LocationEngine locationEngine,
+                           @NonNull LocationEngineRequest locationEngineRequest,
+                           @NonNull NavigationEventDispatcher navigationEventDispatcher,
+                           @NonNull MapboxNavigator mapboxNavigator,
+                           @NonNull ScheduledExecutorService executorService) {
+    this.locationEngine = locationEngine;
+    this.locationEngineRequest = locationEngineRequest;
+    this.navigationEventDispatcher = navigationEventDispatcher;
+    this.mapboxNavigator = mapboxNavigator;
+    this.executorService = executorService;
+  }
+
+  void configure(@NonNull File path, @NonNull String version, @NonNull String host,
+                 @NonNull String accessToken,
+                 @NonNull OnOfflineTilesConfiguredCallback onOfflineTilesConfiguredCallback) {
+    OfflineNavigator offlineNavigator =
+        new OfflineNavigator(mapboxNavigator.getNavigator(), version, host, accessToken);
+    String tilePath = new File(path, version).getAbsolutePath();
+    offlineNavigator.configure(tilePath, onOfflineTilesConfiguredCallback);
+  }
+
+  void start() {
+    if (future == null) {
+      locationEngine.requestLocationUpdates(locationEngineRequest, callback, null);
+      final Handler handler = new Handler(Looper.getMainLooper());
+      future = executorService.scheduleAtFixedRate(new Runnable() {
+        @Override
+        public void run() {
+          if (rawLocation != null) {
+            final Location enhancedLocation = getLocation(new Date(), 0, rawLocation);
+            handler.post(new Runnable() {
+              @Override
+              public void run() {
+                navigationEventDispatcher.onEnhancedLocationUpdate(enhancedLocation);
+              }
+            });
+          }
+        }
+      }, 1500, 1000, TimeUnit.MILLISECONDS);
+    }
+  }
+
+  void stop() {
+    if (future != null) {
+      stopLocationUpdates();
+    }
+  }
+
+  void kill() {
+    if (future != null) {
+      stopLocationUpdates();
+      executorService.shutdown();
+    }
+  }
+
+  private void stopLocationUpdates() {
+    locationEngine.removeLocationUpdates(callback);
+    future.cancel(false);
+    future = null;
+  }
+
+  private Location getLocation(Date date, long lagMillis, Location rawLocation) {
+    NavigationStatus status = mapboxNavigator.retrieveStatus(date, lagMillis);
+    return getMapMatchedLocation(status, rawLocation);
+  }
+
+  private Location getMapMatchedLocation(NavigationStatus status, Location fallbackLocation) {
```

Comment:
What is the fallback location actually used for? All values are overwritten from the navigation status anyway.